### PR TITLE
Add immediate backward and refined recompute for CSAIndexer

### PR DESCRIPTION
**Pull request by lshpku**
### PR Category
Execute Infrastructure

### PR Types
New features

### Description
对 CSAIndexer 进行立即反向，以及实现 refined recompute 式的输出结果保存

1. <b>立即反向：</b>由于 CSAIndexer 的计算图和主计算图完全分离，它的反向不需要等主干反向的时候才进行，而是可以立即反向，从而释放激活显存

2. <b>RR：</b>在 full recompute 下，CSAIndexer 的前向不需要算两次，只要第一次前向时保存结果，第二次直接从队列里拿即可

Indexer 的输出是一个 [batch_size, q_seq_len, indexer_topk] int32，对于 128K CP8 MBS1 场景，每个输出的大小仅 32MB，保存下来非常划算

### 是否引起精度变化
否

### CODE REVIEW

File: src/paddlefleet/transformer/csa_attention.py [hunk 1]
```diff
@@ -811,102 +812,33 @@ def forward(
...
-                        )
+                    if not framework._dygraph_tracer()._has_grad:
+                        with paddle.enable_grad():
+                            compress_topk_idxs, indexer_loss = (
+                                indexer_forward_func()
+                            )
+                        indexer_loss.backward()
+                        self.indexer_rr_queue.put(compress_topk_idxs)
+                    else:
+                        compress_topk_idxs = self.indexer_rr_queue.get_nowait()
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/csa_attention.py:839]: P1 **优先级：P1**

这里把 `self.training` 下的 `_has_grad=True` 都当成 refined recompute 的第二次 forward，并立即从 `indexer_rr_queue` 取结果。但 `DSv4HybridAttention.forward` 直接调用 `self.core_attention(...)`，普通训练（未开启 full recompute/RR）也会进入训练态的有梯度 forward；此时队列从未在前一次 `not _has_grad` 分支写入，ratio=4 层会在这里抛 `queue.Empty`。现有 `TestDSv4HybridAttentionForwardBackward.test_backward_gradient_flow` 这类非 recompute 训练用例就会覆盖该场景。请不要只用 `_has_grad` 判定 RR 阶段，至少要用外层 full-recompute/RR 配置显式控制；非 RR 训练路径应继续正常计算 indexer 并把 loss 接回，或明确在这里执行一次 immediate backward。

可按下面的结构拆分路径：

```python
use_indexer_rr = self.training and self._indexer_rr_enabled_for_this_layer()
if use_indexer_rr:
    if not framework._dygraph_tracer()._has_grad:
        with paddle.enable_grad():
            compress_topk_idxs, indexer_loss = indexer_forward_func()
        indexer_loss.backward()
        self.indexer_rr_queue.put(compress_topk_idxs)
    else:
        compress_topk_idxs = self.indexer_rr_queue.get_nowait()
else:
    compress_topk_idxs, indexer_loss = indexer_forward_func()
    # 非 RR 路径需要保持 indexer_loss 的梯度接入，或明确在这里执行一次 backward。
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已完成初次代码审查，发现一个会影响普通训练路径的阻塞问题，详细说明已放在行级评论中。当前部分单卡、多卡和集成测试仍在运行或排队，建议修复后再等待 CI 完整通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
